toolchain/check/cpp/type_mapping.cpp

```diff
@@ -39,16 +42,29 @@ static auto FindIntLiteralBitWidth(Context& context, SemIR::InstId arg_id)
       arg_const_id.is_symbolic()) {
     // TODO: Add tests for these cases.
     return IntId::None;
+    ;
   }
   auto arg = context.insts().GetAs<SemIR::IntValue>(
       context.constant_values().GetInstId(arg_const_id));
-  unsigned arg_non_sign_bits =
-      context.ints().Get(arg.int_id).getSignificantBits() - 1;
+  llvm::APInt arg_val = context.ints().Get(arg.int_id);
+  unsigned arg_non_sign_bits = arg_val.getSignificantBits() - 1;
 
-  // TODO: What if the literal is larger than 64 bits? Currently an error is
-  // reported that the int value is too large for type `i64`. Maybe try to fit
-  // in i128/i256? Try unsigned?
-  return (arg_non_sign_bits <= 32) ? IntId::MakeRaw(32) : IntId::MakeRaw(64);
+  // Value doesn't fit to any C++ supported signed integer, diagnose and return
+  // the maximum supported integer bit width.
+  if (arg_non_sign_bits >= 128) {
+    CARBON_DIAGNOSTIC(IntTooLargeForCppType, Error,
+                      "integer value {0} too large to be fitted in any "
+                      "supported signed C++ type, max supported _int128",
```

Comment:
How about something like this:
```suggestion
                      "integer value {0} too large to fit in a signed C++ integer type; requires {1} bits, but max is 128",
```
where `{1}` is `arg_non_sign_bits + 1`

Reply:
sg, done.